### ISSUE TITLE
Use a proper lexer for hiding code in doctests

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#84478]
Right now, rustdoc uses line-by-line heuristics to tell if a line should be hidden. That has false positives with multi-line strings, e.g.
```rust
let s = "hello
# because why not?";
```
and soon, after https://github.com/rust-lang/rust/pull/84445,
```rust
let s = "hello
#because why not?";
```

It would be nice to use a proper lexer so we can detect where strings start and end instead of having to do custom "parsing".

_Originally posted by @GuillaumeGomez in https://github.com/rust-lang/rust/issues/84445#issuecomment-825659604_

**Comment by SergioBenitez** [comment]
Please note that this is a breaking change. What's more, this change makes it impossible to hide a multi-line string in a doc comment:

```rust
# let s = "hello\
# there";
```

Please see my suggestion in https://github.com/rust-lang/rust/issues/84502#issuecomment-826031155 which doesn't suffer from this issue.

**Comment by jyn514** [comment]
I don't think this is important enough to spend time on an RFC.
